src/lsp/main.ts

```diff
@@ -160,13 +160,8 @@ function createClient(clojureLspPath: string, fallbackFolder: FallbackFolder): L
             return next(document, position, token);
           }
         },
-        async provideCompletionItem(document, position, context, token, next) {
-          const items = await provideCompletionItems(document, position, token, context);
-          if (items) {
-            return null;
-          } else {
-            return next(document, position, context, token);
-          }
+        provideCompletionItem(document, position, context, token, next) {
+          return null;
```

Comment:
From what I gather:

- returning null in this middleware means the lsp client, and thus calva, will not proceed to procure clojure-lsp's default completion item candidates

If we introduce your other changes but not this one, what would happen? Would the merged completions then be overwritten by clojure-lsp's outright?

Reply:
There's basically two ways in vscode to get completions: have a CompletionProvider or have middleware. We had both, so every completion was actually run twice.
I could also have deleted the provider and put the code into the middleware, but the provider feels like a cleaner solution.

And to answer your question: I think what ever returns first gets used, so not ideal.

Reply:
Nice. Yeah makes sense. Was interesting learning about [the two ways you speak of](https://code.visualstudio.com/api/language-extensions/programmatic-language-features). I'm surprised to have found little about the semantics/consequences of "collisions" here. Still, very excited about the potential here.

Reply:
Yeah, I think clojure/calva is kind of unique in getting its information from two different sources. Usually you'd use the Provider to do the calculations inside the extension if you don't have an LSP or just do everything inside the LSP. Using the middleware for merging is kind of weird, because the `next` call might just cancel your promise :)

Reply:
To the question. I think it might end up with two answers and the user would be offered a choice. A rather confusing choice where the options often would be identical. Not sure about this, but I think I have seen it happen few time when hacking on these things.

> the provider feels like a cleaner solution

Agree.

Reply:
@riotrah btw you'll actually find a better explanation of the different ways to handle this [here](https://code.visualstudio.com/api/language-extensions/embedded-languages)